pkg/metrics/metrics.go

```diff
@@ -0,0 +1,67 @@
+/*
+Copyright 2018 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package metrics
+
+import (
+	"github.com/prometheus/client_golang/prometheus"
+	dto "github.com/prometheus/client_model/go"
+
+	"k8s.io/kube-state-metrics/pkg/options"
+)
+
+type GathererFunc func() ([]*dto.MetricFamily, error)
+
+func (f GathererFunc) Gather() ([]*dto.MetricFamily, error) {
+	return f()
+}
+
+func FilteredGatherer(r prometheus.Gatherer, whitelist options.MetricSet, blacklist options.MetricSet) prometheus.Gatherer {
+	whitelistEnabled := !whitelist.IsEmpty()
+	blacklistEnabled := !blacklist.IsEmpty()
+	filterEnabled := whitelistEnabled || blacklistEnabled
+
+	return GathererFunc(func() ([]*dto.MetricFamily, error) {
+		metricFamilies, err := r.Gather()
+		if err != nil {
+			return nil, err
+		}
+
+		if filterEnabled {
+			newMetricFamilies := []*dto.MetricFamily{}
+			for _, metricFamily := range metricFamilies {
+				// deferencing this string may be a performance bottleneck
+				name := *metricFamily.Name
+				if whitelistEnabled {
+					_, onWhitelist := whitelist[name]
+					if onWhitelist {
+						newMetricFamilies = append(newMetricFamilies, metricFamily)
+					}
+					continue
+				}
+
+				_, onBlacklist := blacklist[name]
```

Comment:
Why do you check whether `whitelistEnabled` is true, but you **don't** check `blacklistEnabled` here.

Reply:
We have to opt for one strategy here. Both the whitelist and blacklist are explicit, so which one is the right one, the one where we explicitly said we definitely want this, or the one where we said we definitely don't want this. As whitelists are usually pretty hand picked I figured that would be the stronger of the two.

Reply:
I'm going to make sure to document he precedence behavior.

Reply:
I changed blacklist and whitelist to be mutually exclusive. It really doesn't make any sense to have both here.